include/albatross/src/models/nearest_neighbor.hpp

```diff
@@ -0,0 +1,115 @@
+
+template <typename FeatureType> struct NearestNeighborFit;
+
+template <typename FeatureType> struct Fit<NearestNeighborFit<FeatureType>> {
+
+  Fit() : training_data(){};
+
+  Fit(const RegressionDataset<FeatureType> &dataset) : training_data(dataset){};
+
+  bool operator==(const Fit<NearestNeighborFit<FeatureType>> &other) const {
+    return training_data == other.training_data;
+  }
+
+  RegressionDataset<FeatureType> training_data;
+};
+
+template <typename DistanceMetric>
+class NearestNeighborModel
+    : public ModelBase<NearestNeighborModel<DistanceMetric>> {
+
+public:
+  NearestNeighborModel() : distance_metric(){};
+
+  std::string get_name() const { return "nearest_neighbor_model"; };
+
+  template <typename FeatureType>
+  Fit<NearestNeighborFit<FeatureType>>
+  _fit_impl(const std::vector<FeatureType> &features,
+            const MarginalDistribution &targets) const {
+    return Fit<NearestNeighborFit<FeatureType>>(
+        RegressionDataset<FeatureType>(features, targets));
+  }
+
+  template <typename FeatureType>
+  auto fit_from_prediction(const std::vector<FeatureType> &features,
+                           const JointDistribution &prediction) const {
+    const NearestNeighborModel<DistanceMetric> m(*this);
+    MarginalDistribution marginal_pred(
+        prediction.mean, prediction.covariance.diagonal().asDiagonal());
+    Fit<NearestNeighborFit<FeatureType>> fit = {
+        RegressionDataset<FeatureType>(features, marginal_pred)};
+    FitModel<NearestNeighborModel, Fit<NearestNeighborFit<FeatureType>>>
+        fit_model(m, fit);
+    return fit_model;
+  }
+
+  template <typename FeatureType>
+  MarginalDistribution
+  _predict_impl(const std::vector<FeatureType> &features,
+                const Fit<NearestNeighborFit<FeatureType>> &fit,
+                PredictTypeIdentity<MarginalDistribution> &&) const {
+    const Eigen::Index n = static_cast<Eigen::Index>(features.size());
+    Eigen::VectorXd mean = Eigen::VectorXd::Zero(n);
+    mean.fill(NAN);
+    Eigen::VectorXd variance = Eigen::VectorXd::Zero(n);
+    variance.fill(NAN);
+
+    for (std::size_t i = 0; i < features.size(); ++i) {
+      const auto min_index =
+          index_with_min_distance(features[i], fit.training_data.features);
+      mean[i] = fit.training_data.targets.mean[min_index];
+      variance[i] = fit.training_data.targets.get_diagonal(min_index);
+    }
+
+    if (fit.training_data.targets.has_covariance()) {
+      return MarginalDistribution(mean, variance.asDiagonal());
+    } else {
+      return MarginalDistribution(mean);
+    }
+  }
+
+private:
+  template <typename FeatureType>
+  std::size_t
+  index_with_min_distance(const FeatureType &ref,
+                          const std::vector<FeatureType> &features) const {
+    assert(features.size() > 0);
+
+    std::size_t min_index = 0;
+    double min_distance = distance_metric(ref, features[0]);
+
+    for (std::size_t i = 1; i < features.size(); ++i) {
```

Comment:
Could we turn `min_distance` into an `optional` so the loop can start at 0? The only difference in the loop would be `!min_distance &&` going at the start of the `if`.

Reply:
I like that pattern better too ... but so far `albatross` doesn't used any optionals!  So we'd have to add a third party lib for it which I've been avoiding (though perhaps the time has come).

Reply:
An alternative is to initialize the `min_distance` to `DBL_MAX` or some such, so it will always be replaced by the first distance.